buildSrc/src/main/java/org/elasticsearch/gradle/testclusters/TestClustersPlugin.java

```diff
@@ -30,53 +31,50 @@
 
-    private static final String LIST_TASK_NAME = "listTestClusters";
     public static final String EXTENSION_NAME = "testClusters";
-    private static final String REGISTRY_EXTENSION_NAME = "testClustersRegistry";
+    public static final String THROTTLE_SERVICE_NAME = "testClustersThrottle";
 
+    private static final String LIST_TASK_NAME = "listTestClusters";
+    private static final String REGISTRY_SERVICE_NAME = "testClustersRegistry";
     private static final Logger logger = Logging.getLogger(TestClustersPlugin.class);
 
-    private ReaperService reaper;
-
     @Override
     public void apply(Project project) {
         project.getPlugins().apply(DistributionDownloadPlugin.class);
-
         project.getRootProject().getPluginManager().apply(ReaperPlugin.class);
-        reaper = project.getRootProject().getExtensions().getByType(ReaperService.class);
+
+        ReaperService reaper = project.getRootProject().getExtensions().getByType(ReaperService.class);
 
         // enable the DSL to describe clusters
-        NamedDomainObjectContainer<ElasticsearchCluster> container = createTestClustersContainerExtension(project);
+        NamedDomainObjectContainer<ElasticsearchCluster> container = createTestClustersContainerExtension(project, reaper);
 
         // provide a task to be able to list defined clusters.
         createListClustersTask(project, container);
 
-        if (project.getRootProject().getExtensions().findByName(REGISTRY_EXTENSION_NAME) == null) {
-            TestClustersRegistry registry = project.getRootProject()
-                .getExtensions()
-                .create(REGISTRY_EXTENSION_NAME, TestClustersRegistry.class);
-
-            // When we know what tasks will run, we claim the clusters of those task to differentiate between clusters
-            // that are defined in the build script and the ones that will actually be used in this invocation of gradle
-            // we use this information to determine when the last task that required the cluster executed so that we can
-            // terminate the cluster right away and free up resources.
-            configureClaimClustersHook(project.getGradle(), registry);
+        // register cluster registry as a global build service
+        project.getGradle().getSharedServices().registerIfAbsent(REGISTRY_SERVICE_NAME, TestClustersRegistry.class, spec -> {});
 
-            // Before each task, we determine if a cluster needs to be started for that task.
-            configureStartClustersHook(project.getGradle(), registry);
+        // register throttle so we only run at most max-workers/2 nodes concurrently
+        project.getGradle()
+            .getSharedServices()
+            .registerIfAbsent(
+                THROTTLE_SERVICE_NAME,
+                TestClustersThrottle.class,
+                spec -> spec.getMaxParallelUsages().set(project.getGradle().getStartParameter().getMaxWorkerCount() / 2)
```

Comment:
I think we set this to half the physical cores in CI, so it may end up lower than it should. 
Looking at some CPU usage graphs will make it  obvious that we are under utilizing.
Unaffected build times are a good sign trough.
 
Since we can't set number of workers in a dynamic way, I think we should keep this code as is 
and stop setting max workers explicitly, I would expect that to work well both for CI and local runs without explicit configuration and with parallel turned on by default.

One caveat for running locally is that we would need to add a memory check 
and error out if there isn't enough.  I have 12 threads and 32GB of ram, and can't run a `--parallel check` without lowering workers. I think we need ~ 3GB for the daemon and ~ an additional 4GB for each worker. The number are based on observations, and the changes in the PR should make them better. 

In theory, worst case, each worker could be running a test  and half as many nodes may be running, each consuming `tests.heap.size` (512MB) by default memory so one needs at lest `daemon memory + 3/4 * nr of workers` GB of memory just to run this. Without taking the OS and OS caches into account.
If we consider that ES recommends leaving half as much memory to OS caches, it would  work out  to 
`daemon memory + number of workers` GB.  We may want to add some margin there, like 16 GB for other stuff running on the machine, that won't hurt CI since we have plenty of memory there. 

On a 12 thread CPU, that works out to 12GB or 15 GB RAM free, so in theory I should be able to run `./graldew --parallel check` without setting the number of workers with this change, given that after running browsers and developer tools I still have  18GB free.

The check may be conditioned on the task graph being big enough so we don't annoy the people to add configuration if they are just looking to run a single test ?
Or maybe that's all just over-complicating it ?

Reply:
> I think we set this to half the physical cores in CI

We use 32 core machines with max-workers set to 16. Are we certain those are 32 _physical_ cores?

> I think we need ~ 3GB for the daemon and ~ an additional 4GB for each worker. 

Our [test workers allocate 512m of ram](https://github.com/elastic/elasticsearch/blob/e34d7fdaf7b511627c64a9e16805fd82f980b8c6/buildSrc/src/main/groovy/org/elasticsearch/gradle/BuildPlugin.groovy#L675). Not sure for the test cluster nodes themselves but running half as many of them in parallel as we currently have the potential for now should help here.

> Or maybe that's all just over-complicating it ?

Maybe not. We could model available memory as a build service just as we do now CPUs with the throttle. It's a more difficult problem though due to irregularities in how much "free" memory is reported by the JVM, other processes running, etc. Gradle tries itself to do this with the daemon and it rarely works properly. I'm apt to see how this works out in practice. I set my max-workers to 6 (physical cores), but as you say, we want to avoid folks having to "tune" their settings.

Reply:
The CI workers have 16 physical cores with hyperthreading AFAIK  
I'm not sure the IO bottleneck will be hit with this change in place, 
but it's certainly possible. 
Test Cluster nodes allocate 512m too.
A memory model would need to account for test runners in addition to ES nodes. 

Enabling parallel without limiting max workers will require some testing, and possibly additional work for sure. 
I think it's fine if we can't account for everything, probably something like producing an error telling people a suggested max workers to put in their config would also be fine ( and we can do that on a some observed CPU to memory ratio ) when running on a non ramdisk. We tune CI separately anyhow and that's unlikely to change, not sure it's worth looking for a one size fits all.